### ISSUE TITLE
Use tiered cache in place of Caffeine cache

### ISSUE THREAD

**Issue by pditommaso** [seqeralabs/wave#775]
[Tiered cache](https://github.com/seqeralabs/wave/issues/594) provides twofolds benefits: 1) the cache content is shared across all replicas, and 2) the cache content is should not be reloaded in across restarts. 

The goal of this issue is to replace the local Caffeine cache with a tiered cache for the following services: 

- [x] [AwsEcrService](https://github.com/seqeralabs/wave/blob/d5c086dac3a41b1a3a4260ae84f980e58b579ace/src/main/groovy/io/seqera/wave/service/aws/AwsEcrService.groovy#L86-L86)
- [x] [RegistryLookupService](https://github.com/seqeralabs/wave/blob/d5c086dac3a41b1a3a4260ae84f980e58b579ace/src/main/groovy/io/seqera/wave/auth/RegistryLookupServiceImpl.groovy#L84-L85)
- [x] [RegistryAuthService](https://github.com/seqeralabs/wave/blob/d5c086dac3a41b1a3a4260ae84f980e58b579ace/src/main/groovy/io/seqera/wave/auth/RegistryAuthServiceImpl.groovy#L61-L61)

Supporting those service require however extending the current tiered cache API adding the ability to use Java object a cache keys other than plain strings as it is now, see here

[abstract class AbstractTieredCache<V extends MoshiExchange> implements TieredCache<String,V> ](https://github.com/seqeralabs/wave/blob/d5c086dac3a41b1a3a4260ae84f980e58b579ace/src/main/groovy/io/seqera/wave/store/cache/AbstractTieredCache.groovy#L48-L48)

This could be achieved introducing an interface like the following one 

```
interface TieredCacheKey {

  String stableHash()

}

```

The concrete key class should take care of proving an implementation for `stableHash`, likely using [sipHash](https://github.com/seqeralabs/wave/blob/d5c086dac3a41b1a3a4260ae84f980e58b579ace/src/main/groovy/io/seqera/wave/util/RegHelper.groovy#L190-L190) algorithm.

**Comment by munishchouhan** [comment]
@pditommaso does tiercache also addresses https://github.com/seqeralabs/wave/issues/747?
because in AwsEcrService we are using cache.synchronous() to avoid concurreny issue

**Comment by pditommaso** [comment]
No, it uses always Ceffeine behind the scene. We need Java 24 to revert that change

**Comment by munishchouhan** [comment]
Since now we need to use string as key, there is an issue with loader
https://github.com/seqeralabs/wave/blob/1e13c9f8f3401f36f1a0fff32c8c41493bbb009e/src/main/groovy/io/seqera/wave/service/aws/AwsEcrService.groovy#L88-L93

here loader is using key (AwsCreds) to compute the value, changing creds to string will not let us compute it anymore.

**Comment by munishchouhan** [comment]
we can store awscreds as it is and generate token, when fetchs from cache

**Comment by pditommaso** [comment]
`AwsCreds` should implement `TieredCacheKey` as I was suggesting, then should be the same

**Comment by munishchouhan** [comment]
yes, that i did here:
https://github.com/seqeralabs/wave/blob/1e13c9f8f3401f36f1a0fff32c8c41493bbb009e/src/main/groovy/io/seqera/wave/service/aws/AwsEcrService.groovy#L69-L77
but loader requires AwsCreds and not a string because it needs compute token from it

**Comment by pditommaso** [comment]
This needs to be changed to take `TieredCacheKey` as key instead of string

https://github.com/seqeralabs/wave/blob/67f91c8b2ba7611eccf55cf45da5256e0465c59e/src/main/groovy/io/seqera/wave/store/cache/AbstractTieredCache.groovy#L48-L48

ei. 


```
abstract class AbstractTieredCache<K extends TieredCacheKey, V extends MoshiExchange> implements TieredCache<K,V> {

  .. 
}


```


make sense ?

**Comment by munishchouhan** [comment]
ok understood, thanks

**Comment by munishchouhan** [comment]
Another issue is Caffeine cache loader is not a Funtion<>, it doesn't have apply Method

**Comment by pditommaso** [comment]
It should not impact, the trick is to replace in the internal implementation the string key with `TieredCacheKey.stableHash`
